api/src/main/java/io/opentelemetry/common/Attributes.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import static io.opentelemetry.internal.Utils.checkArgument;
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.TreeMap;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * An immutable container for attributes. The type parameter denotes the type of the values of the
+ * attributes.
+ */
+@Immutable
+public abstract class Attributes<T> {
```

Comment:
Can't an attribute map have values of different types? Or is that only for resource?

Reply:
For reference here's the class in the collector, for consistency it seems to make sense to have a similar API with the mutators on a builder and accessors on the object

https://github.com/open-telemetry/opentelemetry-collector/blob/master/consumer/pdata/common.go#L311

Reply:
The variant types for values are encapsulated by the AttributeValue class, so you can declare `Attributes<AttributeValue>` if you need to have the values vary in type. That will be the most common use-case, but for metric labels, we'll want to use `Attributes<String>`.

Remember, the use case here is API surface, so the shape of the API won't necessarily match that in the collector, which doesn't expose an API surface for instrumentation.

Reply:
Thanks for sharing the examples. I agree that the API often won't match, but for general data containers I figure the principles of creating an easy-to-use API are shared to a large degree. Anyways the collector is mainly a reference point for a similar API.

Some observations from my own usage of these APIs
- Wrapping all operations with `stringAttributeValue` in the Java SDK is tedious and significantly reduces the readability of the code
- Similar manipulation of attributes in the collector was pleasant, I was very happy to see the type-specific mutators
- Most manipulation in the collector was unit tests. I presume unit tests in the SDK will similarly create these data containers a lot, so I would argue that increases the potential benefit in a user-friendly API

I'd generally recommend imagining what user code looks like when using an API when creating it and working backwards from there. I can imagine these

```java
Attributes attributes = Attributes.newBuilder()
  .putString("key1", "value1")
  .putInt("key2", 3)
  .build();
```

vs

```
Attributes<AttributeValue> attributes = Attributes.of(
  "key1", stringAttributeValue("value1"),
  "key2", intAttributeValue(3));
```

and would argue the contents of this latter are significantly obfuscated by the wrappers.

My suggestion would be that having a generic class as a backing store makes sense, and the actual user API could be type-specific, so `Attributes` and `Labels` or something, first is polymorphic second is only String. Former could wrap in `attributeValue` transparent to the user.

Anyways my 2c and sharing some experience with using the current APIs. Hope it helps.

Reply:
Some of us have actually have proposed an alternative API for this that I think is a bit more fluent for real-life instrumentation use-cases, and saves some allocations in the normal use as well. You can see some of the ideas over in this draft PR:  #1130. In particular, moving the type safety to the key-side of the API I think makes it easier to pre-declare the attribute keys with wrappers, and allows the values to be non-wrapped.

Personally, I find the AttributeValue to be very cumbersome for the most common use-cases, as you have also stated.

Reply:
Thanks for sharing - yeah the linked PR's API looks quite nice to me :)